### ISSUE TITLE
Designing a game server using CAF

### ISSUE THREAD

**Issue by ethindp** [actor-framework/actor-framework#1319]
So I'm planning to design a game server for a private project and am considering different application designs. CAF caught my eye as a potentially extremely powerful mechanism, but I'd like to know a couple things first:

1. Can I mix this and an entity-component system (ECS) architecture, like entt?
2. Depending on the above question:
    1. If no, can I override the network IO routines to use my own protocol (e.g. QUIC, JSON-RPC)?
    2. If yes, what would be the most elegant way of doing this?

I don't know how this framework would cope with real-time games (e.g. FPSes), or even if its appropriate, particularly if things like real-time physics (for example) is involved. So, what tips would you guys suggest for evaluating whether the Actor model is a good design model? I'm new to models like the actor model so I'm fascinated by it and find it to be really cool.

**Comment by Neverlord** [comment]
I'm not familiar with entt, but combining CAF with other systems is generally not an issue as long as the other framework does not cause intra-actor concurrency. If the other framework has its own thread pool, then you need to be aware that you are running two separate schedulers and thread pools. CAF allows you to configure the number of threads, but running two thread pools side by side may come with some inefficiencies. There is an interface for the scheduler, so CAF could theoretically integrate into another scheduler. I can't offer any experience on that, though.

Regarding networking: don't use the I/O module if you don't need to. :)

Using CAF actors for concurrency alone is fine and a common use case. We are also working on a new network layer in the incubator (caf_net) that makes interfacing with other network protocols easier. There isn't much documentation available right now, though. Just some examples for WebSocket communication. Just wanted to mention it.Message ID: ***@***.***>

**Comment by ethindp** [comment]
@Neverlord EnTT is an entity component system framework for C++. Why should I avoid the networking module? And how would I structure my program if I did avoid it?

**Comment by Neverlord** [comment]
The I/O module and in particular the function pair publish/remote_actor is designed for in-cluster communication between trusted nodes. Connecting actor systems in this way makes actor communication transparent and gives all nodes the ability to communicate to any other actor in the cluster and even to remotely spawn actors (when registering actor types).

With external clients (players), care must be taken to allow only certain operations and messages. Furthermore, the network protocol spoken by CAF requires you to run the same CAF version among all nodes.

A common design pattern is to structure the application internally with actors, use the I/O module to connect the cluster and rely on standard (or at least well-defined) interfaces at the edges that bridge to external clients.

I didn't mean to advise avoiding the I/O module in general, but I do recommend not exposing actor/CAF interfaces to external parties. Hope that makes it more clear. :)Message ID: ***@***.***>

**Comment by ethindp** [comment]
@Neverlord Ah, okay, so use an external network protocol for players, and then translate that into actor messages internally. I think I understand now.

**Comment by ethindp** [comment]
Actually, this might be best: since the client and server would probably both be using CAF (unless CAF would be bad for game clients), I could just have an actor that handles networking. Create a detached actor running in its own thread, maybe, and I could just send messages back and fourth between it and other actors. When it received a new packet, it would translate that into an actor command/message and send it to an actor to handle it (or if I could get the actor system it could just spawn a new actor for handling). Then for sending packets, the process just happens in reverse: actor sends message to network actor, network actor translates that into packet and sends it to client. Thoughts?

**Comment by Neverlord** [comment]
Yep, I think structuring it this way is good practice. You can either use a detached actor doing the networking, use a broker (I/O actor), or write a custom transport in caf_net (if you want to peek at what's coming to CAF). Integration other networking libraries is also an option as long as you can take care of making sure to not overload your actors (with some form of back pressure, condition variables of what-have-you).Message ID: ***@***.***>
